queue_consumer.go

```diff
@@ -132,30 +151,48 @@ func (mf *Consumer) Run(ctx context.Context) error {
 	del := NewBatchDeleter(cleanupCtx, cleanupWG, mf.s, mf.DeleteMessageAccumulatorTimeout, mf.DeleteMessageDrainTimeout)
 
 	messages := make(chan job)
-	go mf.receiveMessages(cleanupCtx, cleanupWG, messages, del)
-	var done bool
+	go mf.receiveMessages(cleanupCtx, cleanupWG, mf.shutDown, messages, del)
+	defer func() {
+		close(jobs)
+		wg.Wait()
+		cleanupCancel()
+		cleanupWG.Wait()
+	}()
+
 	for {
 		// Stop if the context was cancelled
 		select {
 		case <-ctx.Done():
-			close(jobs)
-			wg.Wait()
-			cleanupCancel()
-			cleanupWG.Wait()
 			return ctx.Err()
-		case msg := <-messages:
-			if done {
-				break
+		case msg, ok := <-messages:
+			if !ok {
+				return nil
 			}
 			select {
 			case <-ctx.Done():
-				done = true
+				return ctx.Err()
 			case jobs <- msg:
 			}
 		}
 	}
 }
 
+// ShutDown initiates the graceful shutdown of the consumer by stopping the ReceiveMessage loop
+// and allowing any in-flight handlers to finish processing.
+//
+// The context can be canceled during this time to abort pending operations early.  This is done co-operatively
+// and requires the provided handler func to honour the context cancelation.
+func (mf *Consumer) ShutDown() {
+	mf.shutDownLock.Lock()
+	defer mf.shutDownLock.Unlock()
```

Comment:
I don't think this mutex is necessary, and it points at a couple potential problems:

1. There should only be one call to `ShutDown()`. If it's being called multiple times, especially concurrently, that suggests logic errors in the caller.
2. If you close an already closed channel (or the nil channel), it will [panic](https://golang.org/ref/spec#Close).

Reply:
I agree on the unnecessary complexity.  My upcoming change will remove this altogether.

As far as the panic, the "lock + select" approach was meant to guard against trying to close the the already closed channel, but it may not be correct, and it's definitely not easy to ready.